spec/src/main/asciidoc/txcontext.asciidoc

```diff
@@ -0,0 +1,193 @@
+        tx.begin();
+        try (Connection con = ds.getConnection()) {
+            return u + con.createStatement().executeUpdate(sql);
+        } finally {
+            tx.commit();
+        }
+    } catch (Exception x) {
+        throw new CompletionException(x);
+    }
+});
+
+tx.begin();
+... do transactional work here
+
+updateCount = stage2.join(); // <-- stage2's action is guaranteed to never
+                             //     run under this transaction because absence
+                             //     of a transaction is propagated to it
+
+... more transactional work
+----
+
+It should be noted that cleared, rather than propagated, transaction context can accomplish the same.
+
+A `ThreadContextProvider` that supports propagation of the absence of a transaction, but not propagation of an active transaction is allowed to raise `IllegalStateException` from its `currentContext` method. The exception flows back to the application on operations such as `managedExecutor.supplyAsync(supplier)`, `threadContext.withContextCapture`, or `threadContext.contextualFunction`, indicating the restriction against propagating active transactions. The `IllegalStateException` should have a meaningful message making it clear to the user that lack of support for the propagation of active transactions is the cause of the error.
+
+For example, the application can expect to see `IllegalStateException` here if the optional behavior of propagating active transactions to other threads is not supported,
+
+[source, java]
+----
+tx.begin();
+stage = executor.runAsync(action); // <-- raises IllegalStateException
+...
+----
+
+=== Propagation of Active Transactions for Serial Use, but not Parallel
+
+Some transaction managers and transactional resources may allow for propagation of an active transaction to multiple threads serially, with the limitation that the transaction is active on at most one thread at any given time.
+
+For example,
+
+[source, java]
+----
+executor = ManagedExecutor.builder()
+           .propagated(ThreadContext.TRANSACTION)
+           .build();
+
+// Allowed because it limits the transaction to serial use
+stage1 = executor.newIncompleteFuture();
+tx.begin();
+try {
+    stage3 = stage1.thenApply(updateCount -> {
+        try (Connection con = dataSource.getConnection()) {
+            return updateCount + con.createStatement().executeUpdate(sql2);
+        } catch (SQLException x) {
+            throw new CompletionException(x);
+        }
+    }).thenApply(updateCount -> {
+        try (Connection con = dataSource.getConnection()) {
+            return updateCount + con.createStatement().executeUpdate(sql3);
+        } catch (SQLException x) {
+            throw new CompletionException(x);
+        }
+    }).whenComplete((result, failure) -> {
+        try {
+            if (failure == null && tx.getStatus() == Status.STATUS_ACTIVE)
+                tx.commit();
```

Comment:
So this executor thread has finished the transaction which is still associated with the thread that called tx_begin on line 102. This means that that thread (the one that created the transaction) is now associated with an invalid transaction.

In the PR #132 I documented best practice I added the following warning:
{quote}
The issue with ending the transaction in a completion stage is that
the context of the initiating thread will be re-applied after the executor finishes and if a stage
ended the transaction then this initiating thread will be associated with the finished transaction
which would mean that that thread could not begin any new transactions.
{quote}

Reply:
That's a good point - it looks like I've missed out on maintaining consistency with 132 in this area.
Although I do wonder if such a limitation is going to be too great of an impediment to anyone wanting to use transaction context propagation in reactive patterns.  And also, we don't seem to have that limitation in our own implementation of transaction context propagation (which is all experimental at this point).  I remember that I did need to modify a few lines of code to check that the transaction is still active before restoring it the thread, but we've never had problems with subsequent transactions being disallowed on the thread after that.

From the user's standpoint, it certainly seems the natural thing to let the chain of dependent stages resolve the transaction.  I wonder if this needs to be another of the optional patterns for transaction context propagation.

Reply:
I just added a commit that explains the above as another optional behavior, along with the exception to expect if unsupported, and includes an example of committing from the main thread instead.  Let me know if this works.  Also, it might make sense to put the changes from 132 into txcontext.asciidoc rather than the general examples asciidoc now that we have a good place for it.  I'm really sorry that pull has been held up so long.

Reply:
Okay, I think you are correct. I will update our SmallRye implementation to perform the same check that you are making, namely to check that the transaction is still active before deciding whether or not to restore it to the thread, thanks. I will also need to update the guidelines I wrote on the PR 132 (for issue #116).

Reply:
@njr-11 In fact I think that the doc updates in this PR are more complete than mine so I would not object if you were to close my PR.

Reply:
I'll go ahead and close it out for now. Feel free to reopen if there ends up being anything else to add.  Thanks for all of your collaboration in this area. It has been very helpful.